src/snowflake/snowpark/_internal/analyzer/select_statement.py

```diff
@@ -324,12 +324,16 @@ def plan_state(self) -> Dict[PlanState, Any]:
 
     @property
     def cumulative_node_complexity(self) -> Dict[PlanNodeCategory, int]:
-        if self._cumulative_node_complexity is None:
-            self._cumulative_node_complexity = sum_node_complexities(
-                self.individual_node_complexity,
-                *(node.cumulative_node_complexity for node in self.children_plan_nodes),
-            )
-        return self._cumulative_node_complexity
+        with self.analyzer.session._plan_lock:
+            if self._cumulative_node_complexity is None:
+                self._cumulative_node_complexity = sum_node_complexities(
+                    self.individual_node_complexity,
+                    *(
+                        node.cumulative_node_complexity
```

Comment:
we might also need to handle encoded_node_id_with_query

Reply:
also _sql_query

Reply:
I can add it for `encoded_node_id_with_query` but sql_query is populated when we call `resolve()`. That should be safe right?

Reply:
also, `encoded_node_id_with_query` is a cached_property which is eagerly executed for `SnowflakePlan` which will also trigger it for `Selectable` during `resolve()`

Reply:
the encoded_node_id_with_query is only used for cte optimization, so it may not be triggered during resolve